ocs_ci/deployment/aws.py

```diff
@@ -134,18 +150,241 @@ def destroy_cluster(self, log_level="DEBUG"):
-            logger.debug(f"Finding volumes with pattern: {volume_pattern}")
-            volumes = self.aws.get_volumes_by_name_pattern(volume_pattern)
-            logger.debug(f"Found volumes: \n {volumes}")
-            for volume in volumes:
-                self.aws.detach_and_delete_volume(
-                    self.aws.ec2_resource.Volume(volume['id'])
+
+class AWSUPI(AWSBase):
+    """
+    A class to handle AWS UPI specific deployment
+    """
+    def __init__(self):
+        self.name = self.__class__.__name__
+        super(AWSUPI, self).__init__()
+
+    class OCPDeployment(BaseOCPDeployment):
+        def __init__(self):
+            super(AWSUPI.OCPDeployment, self).__init__()
+            upi_repo_name = f'openshift-misc-{config.RUN["run_id"]}'
+            self.upi_repo_path = os.path.join(
+                constants.EXTERNAL_DIR, upi_repo_name,
+            )
+
+            self.upi_script_path = os.path.join(
+                self.upi_repo_path,
+                'v3-launch-templates/functionality-testing'
+                '/aos-4_2/hosts/'
+            )
+
+        def deploy_prereq(self):
+            """
+            Overriding deploy_prereq from parent. Perform all necessary
+            prerequisites for AWSUPI here.
+            """
+            super(AWSUPI.OCPDeployment, self).deploy_prereq()
+
+            # setup necessary env variables
+            upi_env_vars = {
+                'INSTANCE_NAME_PREFIX': config.ENV_DATA['cluster_name'],
+                'AWS_REGION': config.ENV_DATA['region'],
+                'rhcos_ami': config.ENV_DATA['rhcos_ami'],
+                'route53_domain_name': config.ENV_DATA['base_domain'],
+                'vm_type_masters': config.ENV_DATA['master_instance_type'],
+                'vm_type_workers': config.ENV_DATA['worker_instance_type'],
+                'num_workers': str(config.ENV_DATA['worker_replicas'])
+            }
+            for key, value in upi_env_vars.items():
+                os.environ[key] = value
+
+            # ensure environment variables have been set correctly
+            for key, value in upi_env_vars.items():
+                assert os.getenv(key) == value, f"{os.getenv(key)} != {value}"
+
+            # git clone repo from openshift-qe repo
+            clone_repo(
+                constants.OCP_QE_MISC_REPO, self.upi_repo_path
+            )
+
+            # Sym link install-dir to cluster_path
+            install_dir = os.path.join(self.upi_script_path, "install-dir")
+            absolute_cluster_path = os.path.abspath(self.cluster_path)
+            logger.info(
+                "Sym linking %s to %s", install_dir, absolute_cluster_path
+            )
+            os.symlink(absolute_cluster_path, install_dir)
+
+            # NOT A CLEAN APPROACH: copy openshift-install and oc binary to
+            # script path because upi script expects it to be present in
+            # script dir
+            bindir = os.path.join(os.getcwd(), 'bin')
```

Comment:
The `bindir` shouldn't be taken from current working directory (`os.getcwd()`) but from configuration parameter `RUN['bin_dir']`. Something like this:

`bin_dir = os.path.abspath(os.path.expanduser(config.RUN['bin_dir']))`

The `os.path.expanduser(...)` is important here, because on Jenkins it is set to:
```
RUN:
  bin_dir: "~/bin"
```

Reply:
+1 Yep, this one approach was mentioned already in some of my previous comment, maybe it was missed. But if I remember correctly I suggested to use this approach directly in config itself so everywhere we will read `RUN['bin_dir']` we will read already path which will be absolute with expanded user.

Reply:
@petr-balogh Agreed, I think it's best if any paths coming from our config are absolute paths and not relative. This would solve issues like this from arising when we are doing something different in jenkins vs. local environments.

For now though, this has been updated to @dahorak 's suggestion.